methods/SPIA/SPIA.py

```diff
@@ -0,0 +1,292 @@
+        self.beta = beta
+        self.markdown = markdown
+        if markdown:
+            if os.path.exists(markdown if '.md' in markdown else markdown.split('.')[0] + '.md'):
+                print("Warning: '" + markdown + "' file already exists and will be overwritten!")
+
+    @staticmethod
+    def load_data_dict(dictionary, all):
+        '''
+
+        Part of this code is imported from https://github.com/iseekwonderful/PyPathway on MIT license.
+        Load data from certain json.
+
+        Args:
+            dictionary: the json, this method will load data form json
+            all: a python list of total genes. e.g. ['A', 'B', 'C', 'D']
+
+        Returns: the loaded data
+
+        '''
+        data = dictionary
+        Bb = set(all)
+        datpT = {}
+        id2name = data['id2name']
+        for pid, v in data.items():
+            if pid == 'id2name':
+                continue
+            datpT[pid] = {}
+            for key, d in v.items():
+                if key == 'row_names':
+                    for idname in range(len(d)):
+                        Aaa = set(d[idname].split(','))
+                        Ddd = list(Aaa.intersection(Bb))
+                        if Ddd:
+                            if len(Ddd) > 1:
+                                print("Error with gene names in pathway occured")
+                            else:
+                                d[idname] = Ddd[0]
+                    datpT[pid][key] = d
+                else:
+                    m = np.zeros(
+                        (len(v['row_names']), len(v['row_names'])))
+                    for x in d:
+                        m[x[0]][
+                            x[1]] = 1
+                    datpT[pid][
+                        key] = m
+        return datpT, id2name
+
+    @staticmethod
+    def calculate_spia(de, all, dictionary, nB=2000, beta=None, combine='fisher'):
+        """
+        This code is imported from https://github.com/iseekwonderful/PyPathway on MIT license.
+        Contains SPIA algorithm.
+
+        Args:
+            de: a python dict of DEGs. key:gene, value:fold-change. e.g. {'A':2.1, 'B':3.3 ...}
+            all: a python list of total genes. e.g. ['A', 'B', 'C', 'D']
+            dictionary: a json- contains data from pathways
+            nB: number of iterations of random sample choosing
+            beta: list of gene relations values, if None the values are default
+            combine: way of calculating pG
+
+        Returns: an array with pathway id, pathway name, pNDE, pPERT, pG, FDR correction,
+        Bonferroni correction, status for each pathway
+
+        """
+
+        de = {k: float(v) for k, v in de.items()}
+        all = [x for x in all]
+        datpT_ALL, id2name = SPIA.load_data_dict(dictionary, all)
+        inter_value = i_val or beta
+        rel_dict = {rel[i]: inter_value[i] for i in range(len(rel))}
+        datp_ALL = {}
+        for k, v in datpT_ALL.items():
```

Comment:
No comments, single-letter variable names - it is difficult to understand this part of the method.